### ISSUE TITLE
Startup failed: cannot pickle user object

### ISSUE THREAD

**Issue by xmcp** [ht35268/bzShare#16]
```python
root@ubuntu-server:/home/xmcp/bzshare# python3 bzshare.py
Traceback (most recent call last):
  File "bzshare.py", line 2, in <module>
    import bzs.core
  File "/home/xmcp/bzshare/bzs/core.py", line 19, in <module>
    from . import module_files
  File "/home/xmcp/bzshare/bzs/module_files.py", line 13, in <module>
    from . import sqlfs
  File "/home/xmcp/bzshare/bzs/sqlfs/__init__.py", line 4, in <module>
    from . import file_system_permissions
  File "/home/xmcp/bzshare/bzs/sqlfs/file_system_permissions.py", line 4, in <mo dule>
    from .. import users
  File "/home/xmcp/bzshare/bzs/users.py", line 150, in <module>
    database = db.Database
  File "/home/xmcp/bzshare/bzs/users.py", line 86, in __init__
    n_usr.save_data()
  File "/home/xmcp/bzshare/bzs/users.py", line 36, in save_data
    bin_data = pickle.dumps(self)
_pickle.PicklingError: Can't pickle <class 'bzs.users.UserManagerType.User'>: at tribute lookup User on bzs.users failed
root@ubuntu-server:/home/xmcp/bzshare#
```

**Comment by xmcp** [comment]
according to [the docs](https://docs.python.org/3.5/library/pickle.html#what-can-be-pickled-and-unpickled), only "classes that are defined at the top level of a module" are picklable.

so maybe you need to put the `User` to the top level of `users.py`.

**Comment by ht35268** [comment]
No.

Your problem is not attributed correctly.

After I released this commit I realized that the real problem is not this. The actual problem is more close to infinite recursion. Using copy.deepcopy() causes endless recursion until it reaches something like:

```Python
Traceback (most recent call last):
  File "bzshare_sqlfs.py", line 13, in <module>
    sqlfs.Filesystem.shell()
  File "D:\Development\Github\bzShare\bzs\sqlfs\file_system.py", line 857, in shell
    ret_result = self.__shell()
  File "D:\Development\Github\bzShare\bzs\sqlfs\file_system.py", line 717, in __shell
    self.copy(src, cmd[2])
  File "D:\Development\Github\bzShare\bzs\sqlfs\file_system.py", line 757, in copy
    ret_result = self.__copy(source, target_parent, new_owners=new_owners)
  File "D:\Development\Github\bzShare\bzs\sqlfs\file_system.py", line 523, in __copy
    target = copy.deepcopy(source)
  File "C:\Programs\Python3\lib\copy.py", line 182, in deepcopy
    y = _reconstruct(x, rv, 1, memo)
  File "C:\Programs\Python3\lib\copy.py", line 298, in _reconstruct
    state = deepcopy(state, memo)
  File "C:\Programs\Python3\lib\copy.py", line 155, in deepcopy
    y = copier(x, memo)
  File "C:\Programs\Python3\lib\copy.py", line 244, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "C:\Programs\Python3\lib\copy.py", line 182, in deepcopy
    y = _reconstruct(x, rv, 1, memo)
  File "C:\Programs\Python3\lib\copy.py", line 292, in _reconstruct
    args = deepcopy(args, memo)
  File "C:\Programs\Python3\lib\copy.py", line 155, in deepcopy
    y = copier(x, memo)
  File "C:\Programs\Python3\lib\copy.py", line 224, in _deepcopy_tuple
    y = [deepcopy(a, memo) for a in x]
  File "C:\Programs\Python3\lib\copy.py", line 224, in <listcomp>
    y = [deepcopy(a, memo) for a in x]
  File "C:\Programs\Python3\lib\copy.py", line 155, in deepcopy
    y = copier(x, memo)
  File "C:\Programs\Python3\lib\copy.py", line 219, in _deepcopy_list
    y.append(deepcopy(a, memo))
  File "C:\Programs\Python3\lib\copy.py", line 182, in deepcopy
    y = _reconstruct(x, rv, 1, memo)
  File "C:\Programs\Python3\lib\copy.py", line 298, in _reconstruct
    state = deepcopy(state, memo)
  File "C:\Programs\Python3\lib\copy.py", line 155, in deepcopy
    y = copier(x, memo)
  File "C:\Programs\Python3\lib\copy.py", line 244, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "C:\Programs\Python3\lib\copy.py", line 182, in deepcopy
    y = _reconstruct(x, rv, 1, memo)
  File "C:\Programs\Python3\lib\copy.py", line 298, in _reconstruct
    state = deepcopy(state, memo)
  File "C:\Programs\Python3\lib\copy.py", line 155, in deepcopy
    y = copier(x, memo)
  File "C:\Programs\Python3\lib\copy.py", line 244, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "C:\Programs\Python3\lib\copy.py", line 182, in deepcopy
    y = _reconstruct(x, rv, 1, memo)
  File "C:\Programs\Python3\lib\copy.py", line 298, in _reconstruct
    state = deepcopy(state, memo)
  File "C:\Programs\Python3\lib\copy.py", line 155, in deepcopy
    y = copier(x, memo)
  File "C:\Programs\Python3\lib\copy.py", line 244, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "C:\Programs\Python3\lib\copy.py", line 182, in deepcopy
    y = _reconstruct(x, rv, 1, memo)
  File "C:\Programs\Python3\lib\copy.py", line 307, in _reconstruct
    y.__dict__.update(state)
AttributeError: 'NoneType' object has no attribute 'update'
```

After this it should be noticed that I should write a copy function of my own. During the former editions of bzShare this did not arose as a problem, and I detected that parental functions would not be copied, yet deepcopy also ignores files, which should be related to another issue.

I am currently forming another copy function, which directly accesses the functions. This may look like reinventing a wheel, but I consider it necessary.

**Comment by ht35268** [comment]
Please give me a pull request or inform <del>where should I put it</del>

Does it mean that `class User` should be independent of `class UserManager` and in the root of `users.py`?

**Comment by ht35268** [comment]
**P.S.** Please capitalize titles.

**Comment by ht35268** [comment]
This might not be fixed until we introduce a fix to #20.

**Comment by ht35268** [comment]
The fix to #20 is not needed.

Please upgrade to Python 3.5.2!
Please upgrade to Python 3.5.2!
Please upgrade to Python 3.5.2!

Important things must be attributed twice.
